src/passes/SignaturePruning.cpp

```diff
@@ -0,0 +1,192 @@
+        info.callRefs = std::move(FindAll<CallRef>(func->body).list);
+        info.usedParams = ParamUtils::getUsedParams(func);
+      });
+
+    // A map of types to all the information combined over all the functions
+    // with that type.
+    std::unordered_map<HeapType, Info> allInfo;
+
+    // Map heap types to all functions with that type.
+    std::unordered_map<HeapType, std::vector<Function*>> sigFuncs;
+
+    // Combine all the information we gathered into that map.
+    for (auto& [func, info] : analysis.map) {
+      // For direct calls, add each call to the type of the function being
+      // called.
+      for (auto* call : info.calls) {
+        allInfo[module->getFunction(call->target)->type].calls.push_back(call);
+      }
+
+      // For indirect calls, add each call_ref to the type the call_ref uses.
+      for (auto* callRef : info.callRefs) {
+        auto calledType = callRef->target->type;
+        if (calledType != Type::unreachable) {
+          allInfo[calledType.getHeapType()].callRefs.push_back(callRef);
+        }
+      }
+
+      // A parameter used in this function is used in the heap type - just one
+      // function is enough to prevent the parameter from being removed.
+      auto& allUsedParams = allInfo[func->type].usedParams;
+      for (auto index : info.usedParams) {
+        allUsedParams.insert(index);
+      }
+      sigFuncs[func->type].push_back(func);
+    }
+
+    // Find parameters to prune.
+    std::unordered_set<HeapType> seen;
+    for (auto& func : module->functions) {
+      auto type = func->type;
+      if (!seen.insert(type).second) {
+        continue;
+      }
+      auto& funcs = sigFuncs[type];
+      if (funcs.empty()) {
+        // We saw no useful functions for this type (this can happen if the type
+        // only appears in an imported function, for example).
+        continue;
+      }
+
+      auto sig = type.getSignature();
+      auto& info = allInfo[type];
+      auto numParams = sig.params.size();
+      if (info.usedParams.size() == numParams) {
+        // All parameters are used, give up on this one.
+        continue;
+      }
+
+      // We found possible work! Find the specific params that are unused try to
+      // prune them.
+      SortedVector unusedParams;
+      for (Index i = 0; i < numParams; i++) {
+        if (info.usedParams.count(i) == 0) {
+          unusedParams.insert(i);
+        }
+      }
+
+      auto oldParams = sig.params;
+      auto removedIndexes = ParamUtils::removeParameters(
+        funcs, unusedParams, info.calls, info.callRefs, module, runner);
+      if (removedIndexes.empty()) {
+        continue;
+      }
+
+      // Success! Update the types.
+      std::vector<Type> newParams;
+      for (Index i = 0; i < numParams; i++) {
+        if (!removedIndexes.has(i)) {
+          newParams.push_back(oldParams[i]);
+        }
+      }
+      auto newType = Signature(Type(newParams), sig.results);
```

Comment:
I believe this is technically unsound because it can optimize two distinct function types to be the same type, which could change the program semantics if it ever tries to cast between the two types. A fix would be to use a TypeBuilder of size 1 to create the new signature type.

Reply:
Ah, the default code path canonicalizes signatures, is that the issue? That rings a bell.

I added a test for it, but it seems to pass even before making this change, see the last 2 commits, so I think I'm missing something...

Reply:
Types are immutable, so it's not possible to modify a signature type in-place. But maybe `newType` here is actually a `Signature` rather than a `HeapType` and it doesn't get turned into a heap type until the type replacement uses a TypeBuilder to do the building?